### PR TITLE
Faceted properties in metadata

### PR DESCRIPTION
**Pull request by chrisdevereux**
This PR adds support for adding _facet_ properties to a collection.

Adding a facet property allows it to build up a hierarchy of records within it. For example:

- A location database may want to define a hierarchy of locations (eg: `Earth > Europe > Berlin`)
- A tagging schema may want to define an ontology where some tags are sub-tags of others (`Animal > Dog > Shiba`)

The facet property type allows these kinds of structures to be built up.

## Demo:

screen

## Details

- A facet property is a reference to a parent record in the same collection. So, in the above location example, setting the parent of a location record 'Europe' to the existing 'Earth' record.
- The system validates against cyclic references in facet properties. So the hierarchy `earth > europe > earth` is not allowed.
- Unlike other property types, facet properties cannot be made either repeated or required.
- When a facet property is imported, the system will attempt to match existing records using `>`-delimited paths, falling back to attempting to match the rightmost path element if that fails.
- The intention here is that when we implement advanced search, it will be possible to search for records via facets - so searching for 'asia' would return records tagged as 'asia', or 'india', or 'mumbai'.
- Similarly, this could be further developed to power a collection browser.

## Limitations & future directions

- Faceted properties are displayed sorted according to their hierarchical structure, but in a flat display. We may in future want to show them in a tree-view or something else that makes the hierarchy clearer.
- Deleting a parent property currently causes child records to become top-level records. A better way of updating related records would be to cascade the delete to the children. Or potentially moving child records to become children the of the deleted record's parent.
- The schema currently does not enforce a maxmum depth. This may be desirable in cases like location where we want a fixed schema for location levels (eg: city > country > district > neighbourhood).
- The schema does not currently enforce a required level of precision for referencing records in hierarchically structured databases. So, in the above example, an editor could chose any of city, district or neighbourhood when defining a location for an asset.

### CODE REVIEW

File: src/app/asset/metadata.entity.ts [hunk 1]
```diff
@@ -353,6 +421,297 @@ export class ControlledDatabaseSchemaPropertyValue
...
+
+      if (res.status !== 'ok') {
+        return ok(undefined);
+      }
+
+      let match: Asset | undefined = res.value.items.find(
+        (x) =>
+          this.getOneFromMetadata<string>(x.metadata) === lastCreatedAsset?.id
+      );
+      if (!match) {
```
Comment (conatus) [inline src/app/asset/metadata.entity.ts:552]: This set of conditionals feels a bit dense and heavily nested. Unsure what one would do with it however. Seems clear enough. If I don't match then I make it.

File: src/frontend/ui/components/schema-editor.component.tsx [hunk 2]
```diff
@@ -432,53 +449,75 @@ export const SchemaEditor: FC<SchemaEditorProps> = ({
...
-            <Box>
-              <Switch
-                sx={{
-                  'input:checked ~ &': {
-                    backgroundColor: 'var(--theme-ui-colors-primary)'
-                  }
-                }}
-                label="Repeated"
-                checked={item.repeated}
+                label="Visible to Public"
```
Comment (conatus) [inline src/frontend/ui/components/schema-editor.component.tsx:497]: Minor nitpick but this copy is a bit misleading. It's more like "Will be included in export".

File: src/test/result.ts [hunk 3]
```diff
@@ -5,7 +5,7 @@ import { ErrorResult, OkResult, Result } from '../common/util/error';
 export function requireSuccess<T>(x: Result<T>) {
   assert(
     x.status === 'ok',
-    'Expected operation to fail. Got error:',
+    'Expected operation to suceed. Got error:',
```
Comment (conatus) [inline src/test/result.ts:8]: Minor typo in the comment: `succeed`.

File: src/app/asset/metadata.entity.ts [hunk 4]
```diff
@@ -1,29 +1,39 @@
 import { Embeddable, Property } from '@mikro-orm/core';
 import { randomUUID } from 'crypto';
 import { isValid, parseISO } from 'date-fns';
-import { keyBy } from 'lodash';
+import { groupBy, keyBy, mapValues } from 'lodash';
```
Comment (conatus) [inline src/app/asset/metadata.entity.ts:4]: Please fix the linting errors.

File: src/app/asset/__tests__/asset.spec.ts [hunk 5]
```diff
@@ -349,6 +355,296 @@ describe(AssetService, () => {
         expect(res).toBeUndefined();
       });
     });
+
+    describe('facets', () => {
```
Comment (conatus) [inline src/app/asset/__tests__/asset.spec.ts:359]: I sort of feel probably this file should be split out. If GitHub is complaining about the size of the diff, the time to split is now!

File: src/app/asset/__tests__/asset.spec.ts [hunk 6]
```diff
@@ -349,6 +355,296 @@ describe(AssetService, () => {
...
+          fixture.archive,
+          fixture.assetCollection.id
+        );
+
+        expect(dbAssets.total).toBe(2);
+        expect(res1).toEqual(referencedAsset.id);
+        expect(res2).toEqual(referencedAsset.id);
+      });
+
+      test('where the target database supports it, facet entries are created on demand', async () => {
```
Comment (conatus) [inline src/app/asset/__tests__/asset.spec.ts:422]: I'd probably be inclined to say something like "the usecase here is tagging". Just to make it clear why you'd want this.

**Review by conatus — CHANGES_REQUESTED** [review]
Nice work! Seems that it nails the ask while keeping the persistance simple, which is very helpful.

In addition there are lots of tests for quite tricky edge cases that keep it thorough. To have an on the demand addition of these (as I understand it) even if this is not yet exposed will make for a nice feature down the line for tagging, as this is the typical behaviour of this sort of feature.

If you could respond to my questions, some for interest and some trying to avoid a couple of things that might trip up in future, that woud be great. Then minor fixes to liner, which VS Code can probably automate and then we can begin QA.

**Comment by conatus** [comment]
Minor point but...we need to do smaller PRs!

File: src/app/asset/__tests__/asset.spec.ts (hunk 6 above)
Comment (conatus) [reply src/app/asset/__tests__/asset.spec.ts:422]: Or put another way, it is impressive it can do this and don't hide the light under a bushel!

File: src/app/asset/__tests__/asset.spec.ts (hunk 5 above)
Comment (chrisdevereux) [reply src/app/asset/__tests__/asset.spec.ts:359]: Yea agree. In general the asset service needs refactoring and splitting out into different services. Would prefer to do this post-earthstar when the functionality has stabilised a bit though!

File: src/app/asset/asset.init.ts [hunk 7]
```diff
@@ -39,7 +39,12 @@ export function initAssets(router: ElectronRouter, media: MediaFileService) {
 
   router.bindArchiveRpc(ListAssets, async (archive, request, range) => {
     return ok(
-      await assetService.listAssets(archive, request.collectionId, range)
+      await assetService.listAssets(
+        archive,
+        request.collectionId,
+        request.sortOptions,
+        range
```
[the remark this answers is not in the document]
Comment (chrisdevereux) [reply src/app/asset/asset.init.ts:46]: It's not exposed in the UI and currently only supports sorting by facet properties - but yes, that's the general idea here.

File: src/app/asset/asset.service.ts [hunk 8]
```diff
@@ -283,27 +285,48 @@ export class AssetService extends EventEmitter<AssetEvents> {
...
+          collections: this.collectionService
+        });
+      } else {
+        entities = await archive.list(
+          AssetEntity,
+          {
+            collection: collectionId
+          },
+          {
+            populate: ['collection', 'mediaFiles'],
```
[the remark this answers is not in the document]
Comment (chrisdevereux) [reply src/app/asset/asset.service.ts:317]: the `populate` option? It's telling the ORM to do a join and also populate the collection and mediaFiles properties when fetching from the database.

File: src/app/asset/asset.service.ts [hunk 9]
```diff
@@ -283,27 +285,48 @@ export class AssetService extends EventEmitter<AssetEvents> {
...
+      }
 
       const result = {
         ...entities,
         items: await this.entitiesToAssets(
           archive,
           entities.items,
           collectionId,
-          { shallow: false }
+          { shallow: false, sort }
```
[the remark this answers is not in the document]
Comment (chrisdevereux) [reply src/app/asset/asset.service.ts:328]: It is:
Screenshot 2022-10-04 at 12 25 05

File: src/app/asset/asset.service.ts [hunk 10]
```diff
@@ -363,10 +391,18 @@ export class AssetService extends EventEmitter<AssetEvents> {
   async filterAssets(
     archive: ArchivePackage,
     collectionId: string,
-    opts: { query: unknown[]; propertyId: string; match: 'any' | 'fuzzy' },
+    opts: {
+      query: unknown[];
```
[the remark this answers is not in the document]
Comment (chrisdevereux) [reply src/app/asset/asset.service.ts:394]: It's not awful (as this is the type expected by the inner api being called to), but better to be more explicit about what types can be used to filter by. Have updated.

File: src/app/asset/asset.service.ts [hunk 11]
```diff
@@ -363,10 +391,18 @@ export class AssetService extends EventEmitter<AssetEvents> {
...
+      propertyId: string;
+      match: 'any' | 'fuzzy';
```
[the remark this answers is not in the document]
Comment (chrisdevereux) [reply src/app/asset/asset.service.ts:396]: Will extract this type to an interface and document the options

File: src/app/asset/metadata.entity.ts [hunk 12]
```diff
@@ -98,7 +110,7 @@ export abstract class SchemaPropertyValue {
   /**
    * True if the property supports multiple occurances.
    */
-  @Property({ type: 'boolean', default: true })
+  @Property({ type: 'boolean', default: false })
```
[the remark this answers is not in the document]
Comment (chrisdevereux) [reply src/app/asset/metadata.entity.ts:110]: Unsure what you mean - are you talking about what the default should be from a user's point of view, or whether it's technically a good idea to give this a default value.

File: src/app/asset/metadata.entity.ts [hunk 13]
```diff
@@ -189,6 +237,25 @@ export abstract class SchemaPropertyValue {
    * Cast to a SchemaProperty instance suitable for returning over APIs
    */
   abstract toJson(): SchemaProperty;
+
+  /** Conveinence from retrieving this property from metadata */
```
[the remark this answers is not in the document]
Comment (chrisdevereux) [reply src/app/asset/metadata.entity.ts:238]: Really need to get a spellcheck action set up on this repo!

File: src/frontend/ui/components/schema-editor.component.tsx (hunk 2 above)
Comment (chrisdevereux) [reply src/frontend/ui/components/schema-editor.component.tsx:497]: Well it's just flat wrong 😅 - good catch!

**Comment by chrisdevereux** [comment]
> Minor point but...we need to do smaller PRs!

yeah..... sorry about this. When we come to do the earthstar integration (after the spike), we should seriously plan how to split that up so as not to make it a nightmare to review.

File: src/app/asset/__tests__/asset.spec.ts (hunk 6 above)
Comment (chrisdevereux) [reply src/app/asset/__tests__/asset.spec.ts:418]: Unsure what you mean - the usecase here is for imports from csv files where the definition needs to be parsed from a string in a csv file, etc.

That should be clearer! Unless I'm missing something here - what do you think it does?

File: src/frontend/ui/components/schema-editor.component.tsx (hunk 2 above)
Comment (chrisdevereux) [reply src/frontend/ui/components/schema-editor.component.tsx:497]: Ah wait - I got confused by the diff formatting.

I don't think that's quite accurate. The current behaviour of this property is that when it is false the property won't be pushed to the website. And perhaps when we have distributed sync, that it won't be pushed to 'external' instances.

It _will_ be included in exports as those are intended to be used to fully recreate the archive (although I guess we might in future want to have an option for more restrictive exports).

I'm not sure 'visible to public' captures this perfectly, but I don't think 'will be included in export' does either...

File: src/app/asset/metadata.entity.ts (hunk 1 above)
Comment (chrisdevereux) [reply src/app/asset/metadata.entity.ts:549]: Gonna quickly refactor.

Comment (chrisdevereux) [reply src/app/asset/metadata.entity.ts]: Better? Added some comments in there and renamed some variables to (hoepfully?) make it a bit easier to follow.

File: src/app/asset/collection.service.ts [hunk 14]
```diff
@@ -225,13 +227,23 @@ export class CollectionService extends EventEmitter<CollectionEvents> {
 
       const schemaDef = schema.map(SchemaPropertyValue.fromJson);
 
+      for (const prop of schemaDef) {
+        if (!prop.canBeRepeated) {
+          prop.repeated = false;
```
[the remark this answers is not in the document]
Comment (chrisdevereux) [reply src/app/asset/collection.service.ts:232]: The functional reason is that unlike other property types, facets can't be required because they need to have one or more roots (that need to be null).

[the remark this answers is not in the document]
Comment (chrisdevereux) [reply src/app/asset/collection.service.ts:232]: The code-level reason is that the repeated & required properties are baked into the base type for properties and I wanted to put some bounds on how much of the code base I was touching with this feature.

I'll open up an issue for refactoring the asset service & metadata. As said elsewhere, I think it needs a bit of a strategic look once we've got earthstar done.

File: src/frontend/ui/components/schema-editor.component.tsx (hunk 2 above)
Comment (conatus) [reply src/frontend/ui/components/schema-editor.component.tsx:497]: I think we can argue about it after people use it. So resolving for now.

File: src/app/asset/__tests__/asset.spec.ts (hunk 6 above)
Comment (conatus) [reply src/app/asset/__tests__/asset.spec.ts:418]: Sorry I thought it was so that if you created a tag, you could create one inline where it didn't exist, like one finds commonly on content management systems. If this is not what it is, which seems to be the case, then we can leave it as is.

**Review by conatus — APPROVED** [review]
